v0.25: Task api (#1208)
meilisearch/documentation: learn/advanced/asynchronous_operations.md

| `finishedAt` | string  | Represents the date and time in the ISO 8601 format when the task has the `failed` or `succeeded` status. The default is `null`.                                                                                                                          |

If a task fails due to an error, all error fields will be appended to the response.

### Task `status`

Task responses always contain a field indicating the request's current `status`. This field can have one of four possible values:

- `enqueued`: the task request has been received and will be processed soon
- `processing`: the task is being processed
- `processed`: the task has been successfully processed
- `failed`: a failure occurred when processing the task

The **final status of a `processed` task is `succeeded` or `failed`.**

### Examples

Suppose you add a new document to your instance using the [documents API endpoint](/reference/api/documents.md#add-or-replace-documents) and receive a `uid`.

When you query the task endpoint using this id, you see that it has been enqueued:

```json
{
    "uid": 1,
    "indexUid": "movies",
    "status": "enqueued",
    "type": "documentsAddition",

>     "details": { "numberOfDocuments": 67493},

Reviewer: Hi @maryamsulemani97 👋 

By default, a details field that can have a real value when the task is failed or succeeded will have a `null` value for `enqueued` and `processing` status.

As for `indexedDocuments` field in this example since the `documentsAddition` is `enqueued`.

>     "details": { "numberOfDocuments": 67493},

Author: Thanks for the update!

While dumps and tasks are both asynchronous operations, they use separate queues and behave differently. For instance, creating a new dump will freeze the task queue until the dump has been generated.

[You can read more about dumps in our dedicated guide.](/reference/features/dumps.md)

## Terminate MeiliSearch while a task is being processed

**Terminating a MeiliSearch instance in the middle of an asynchronous operation is completely safe** and will never adversely affect the database.

MeiliSearch's asynchronous tasks are atomic. This means that all operations concerning a specific task are bundled in one transaction. If any of those operations fails or is interrupted before reaching its end, nothing is committed to the database.

What happens to an asynchronous operation when MeiliSearch is terminated changes depending on the request's `status`:

- `enqueued`: the task will remain enqueued and will be processed as usual
- `processing`: there will be no consequences, since no part of the task has been committed to the database. After restarting, MeiliSearch will treat the task as `enqueued`
- `processed`: there will be no data loss since the request was successfully completed

> - `failed`: the update failed and nothing has been added to the database

Reviewer: ```suggestion
- `failed`: the task failed and nothing has been altered into the database
```

## Which operations are async?

Every operation that might take a long time to be processed is handled asynchronously.

For example, updating the `filterableAttributes` index setting will require as much time as re-indexing all the documents in this index. Because of that, MeiliSearch adds your update request to the task queue and processes it as soon as possible.

Currently, these are MeiliSearch's asynchronous operations:

- Updating index settings
- Adding documents to an index
- Updating documents in an index
- Creating an index
- Updating an index
- Deleting an index
- Updating index settings
- Adding documents to an index
- Updating documents in an index
- Deleting documents from an index
- [Creating a dump](#dumps)

## Understanding tasks

Most of MeiliSearch's asynchronous operations belong to a category called "tasks". After you have requested an asynchronous operation, you can use the [task API](/reference/api/tasks.md) to find the detailed status of your request. To do so, you will need your task's `uid`.

### Response

> The response from the [task API](/reference/api/tasks.md) will always include the following fields in the stated order:

Reviewer: @guimachiavelli, we have control over what comes out of the API directly. What could possibly happen is that a user uses a particular HTTP client that reorders the JSON on its side. In the specification, the defined order of the fields is obviously for the outgoing response from the API.

I don't know how useful it is to specify it in the documentation. 🤓
